sdk/cosmosdb/cosmos/src/retry/timeoutFailoverRetryPolicy.ts

```diff
@@ -0,0 +1,87 @@
+import { RetryPolicy } from "./RetryPolicy";
+import { StatusCodes } from "../common/statusCodes";
+import { GlobalEndpointManager } from "../globalEndpointManager";
+import { HTTPMethod, isReadRequest } from "../common";
+import { ErrorResponse } from "../request";
+import { Constants, ResourceType } from "../common/constants";
+import { RetryContext } from "./RetryContext";
+import { CosmosHeaders } from "../queryExecutionContext/CosmosHeaders";
+import { OperationType } from "@azure/cosmos";
+
+export class TimeoutFailoverRetryPolicy implements RetryPolicy {
+  private maxRetryAttemptCount = 120;
+  private maxServiceUnavailableRetryCount = 1;
+  public retryAfterInMs = 0;
+  public failoverRetryCount = 0;
+  public request: any;
+  public locationEndpoint: any;
+
+  constructor(
+    private globalEndpointManager: GlobalEndpointManager,
+    private headers: CosmosHeaders,
+    private methodType: HTTPMethod,
+    private resourceType: ResourceType,
+    private operationType: OperationType,
+    private enableEndPointDiscovery: boolean
+  ) {}
+
+  private needsRetry(): boolean {
+    const isQuery = Constants.HttpHeaders.IsQuery in this.headers;
+    const isQueryPlan = Constants.HttpHeaders.IsQueryPlan in this.headers;
+    if (this.methodType === HTTPMethod.get || isQuery || isQueryPlan) {
+      return true;
+    }
+    return false;
+  }
+
+  public async shouldRetry(
+    err: ErrorResponse,
+    retryContext?: RetryContext,
+    locationEndpoint?: string
+  ): Promise<boolean> {
+    if (!err) {
+      return false;
+    }
+
+    if (!retryContext || !locationEndpoint) {
+      return false;
+    }
+
+    if (!this.needsRetry()) {
+      return false;
+    }
+
+    if (!this.enableEndPointDiscovery) {
+      return false;
+    }
+
+    if (
+      err.statusCode === StatusCodes.ServiceUnavailable &&
+      this.failoverRetryCount >= this.maxServiceUnavailableRetryCount
+    ) {
+      return false;
+    }
+
+    if (this.failoverRetryCount >= this.maxRetryAttemptCount) {
+      return false;
+    }
+    const canUseMultipleWriteLocations = this.globalEndpointManager.canUseMultipleWriteLocations(
+      this.resourceType,
+      this.operationType
+    );
+    const readRequest = isReadRequest(this.operationType);
+
+    if (!canUseMultipleWriteLocations && !readRequest) {
+      return false;
+    }
+    this.failoverRetryCount++;
+    retryContext.retryCount++;
+
+    if (readRequest) {
```

Comment:
None of the 503 mechanics in .net mark the Endpoint unavailable. 503 or timeouts are not definitive signals that mean a region is unavailable.

Reply:
See https://github.com/Azure/azure-cosmos-dotnet-v3/blob/ee7f86ddd2ea141103c0d57c82e8ff429cd1f6ed/Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs#L395

Reply:
I see endoint getting set unavailable in case of 503 here: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs#L250-L263
and post that retry logic is called.

Reply:
That is only when something called Partition Level Failover is enabled on the account, which was a feature that so far was never done. If you look at the TryMarkEndpointUnavailableForPartitionKeyRange code is actually empty

Reply:
Understood, would update it.

Reply:
@ealsur I have made the changes but I'm having trouble understanding the specific process by which we determine the next preferred location.

Reply:
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/89cc257324d8caf8fe9826c7dfebf88add7ad42c/Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs#L147